community/kernel/src/main/java/org/neo4j/graphdb/Result.java

```diff
@@ -188,4 +188,54 @@
      * @return an iterable of all notifications created when running the query.
      */
     Iterable<Notification> getNotifications();
+
+    /**
+     * Visits all rows in this Result by iterating over them.
+     *
+     * This is an alternative to using the iterator form of Result. Using the visitor is better from a object
+     * creation perspective.
+     *
+     * @param visitor the ResultVisitor instance that will see the results of the visit.
+     */
+    void accept( ResultVisitor visitor );
+
+    /**
+     * Describes a row of a result. The contents of this object is only stable during the
+     * {@linkplain ResultVisitor#visit(ResultRow) visit} call. The data it contains can change between calls to
+     * {@linkplain ResultVisitor#visit(ResultRow) the visit method}. Instances of this type should thus not be saved
+     * for later, or shared with other threads, rather the content should be copied.
+     */
+    interface ResultRow
```

Comment:
Can we have something like a `Set<String> keySet( );` method or alternatively a `Map<String, Object> toMap( );`?. When turning the visitor into an iterator of `Map<String, Object>` we need someway of discovering the used keys.